---
title: "About"
author: "Sam Stroupe"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

#### samstroupe@tamu.edu 

> library(shiny)
runGitHub("ParentageApp", "samstroupe")

The primary purpose of this app is to create a user friendly way to take advantage of R packages for parentage assignment and population evaluation using genomic data. The parentage assignment function of this app uses the R package [Sequoia](https://github.com/JiscaH/sequoia) to determine the best sire and dam for each offspring (Huisman 2017).  The data input is a genotype file in plink.raw format and a csv file with ID, Sex and Birth Year of each sample. [VCFtools](https://vcftools.github.io/man_latest.html) and [Plink](https://www.cog-genomics.org/plink2) can be used to get a vcf file into the correct file format for this app using the follow lines of code in the command line:

> $ vcftools --vcf file.vcf --plink --out fileset   
$ plink --file fileset --recode A --out file_output

Various packages were used to evaluate population genetics data through this app. Many of the packages used were developed by the Grünwald lab. They have a helpful tutorial [here](http://grunwaldlab.github.io/Population_Genetics_in_R/index.html) for explanation on the methods used in this app and more. The input for these packages is a filtered vcf with only biallelic SNPs.

***
The following packages were used to create this app: shiny, sequoia, Rcpp, pedantics, kinship2, ggplot2, vcfR, poppr, ape, RColorBrewer, reshape2, adegenet, cowplot, Cairo, shinyWidgets, grDevices, shinyjs, and shinythemes.

***
#### References:

Huisman, J. (2017). Pedigree reconstruction from SNP data: parentage assignment, sibship clustering and beyond. Molecular ecology resources, 17(5), 1009-1024.  
Niklaus J. Grünwald, Zhian N. Kamvar, Sydney E. Everhart, Javier F. Tabima, and Brian J. Knaus © 2017, Corvallis, Oregon, USA

